C7GameData/Tile.cs

```diff
@@ -55,12 +58,148 @@ public class Tile
+			bool lake = true;
+
+			int i = 0;
+			while(i < contiguous.Count())
+			{
+				Tile currentTile = contiguous[i];
+
+				// Skip land tiles and remove from the stack. This should be checked before they're added but just in case
+				if(!currentTile.IsWater())
+				{
+					currentTile.isLake = 0;
+					contiguous.Remove(currentTile);
+					continue;
+				}
+				contiguousWater += 1;
+				if(contiguousWater > 20)
+				{
+					lake = false;
+					break;
+				}
+
+				// If a contiguous tile is a lake, this (and all contiguous water) is too. This works because contiguousness is transitive. If it is *not* a lake (and is a water tile), then this isn't either.
+				if(currentTile.isLake == 1)
+				{
+					lake = true;
+					break;
+				}
+				if(currentTile.isLake == 0)
+				{
+					lake = false;
+					break;
+				}
+				// If we haven't found out one way or the other, add all of this tile's water neighbors to the stack and then remove this tile
+				foreach((TileDirection dir, Tile neighbor) in currentTile.neighbors)
+				{
+					if(!neighbor.IsWater())
+					{
+						continue;
+					}
+					else if(contiguous.Contains(neighbor) || searchedWater.Contains(neighbor))
+					{
+						continue;
+					}
+					else
+					{
+						// Add neighbor to tiles we should search as long as there's no isthmus blocking
+						if(currentTile.isthmusAt(dir))
+						{
+							continue;
+						}
+						else
+						{
+							contiguous.Add(neighbor);
+						}
+					}
+				}
+				i ++;
+			}
+			if(lake)
+			{
+				// Update all the contiguous water tiles we've looked at. It won't be the whole body of water but it'll save us some repeats.
+				foreach(Tile currentTile in contiguous)
+				{
+					if(currentTile.IsWater())
+					{
+						currentTile.isLake = 1;
+					}
+				}
+				return true;
+			}
+			else
+			{
+				// Update searched water as non-lake.
+				foreach(Tile currentTile in contiguous)
+				{
+					currentTile.isLake = 0;
+				}
+				/*foreach(Tile currentTile in searchedWater)
+				{
+					currentTile.isLake = 0;
+				}*/
+				return false;
+			}
+		}
+
+		public bool providesFreshWater ()
+		{
+			return IsLake() || BordersRiver() || NeighborsLake();
+		}
+
+		// Check for an isthmus that can separate lakes from oceans, affect navigability, etc.
+		// Might want to store this for efficiency? Not sure how important this is in terms of runtime. Should only be run on two water tiles during IsLake
+		public bool isthmusAt (TileDirection dir)
+		{
+			
+			Tile adjacent = neighbors[dir];
+
+			if(adjacent != Tile.NONE && IsWater() && adjacent.IsWater() && dir.isDiagonal())
+			{
+				(TileDirection d1, TileDirection d2) = dir.diagonalComplementary();
+				Tile t1 = neighbors[d1];
+				Tile t2 = neighbors[d2];
+
+				if(!t1.IsWater() && !t2.IsWater())
+				{
+					// An isthmus separates diagonal water tiles with coastal tiles in both complementary directions
+					return true;
+				}
+			}
+			return false;
+		}
```

Comment:
Nicely done.  I doubt this would be called repeatedly during runtime, so I wouldn't go to a lot of extra effort to cache it, but it's never going to change and it would just be storing one variable, so the extra effort isn't a lot.  Your call.